test/features/test_multiscale_feature_persistence.cpp

```diff
@@ -0,0 +1,236 @@
+       0, 0, 0, 7.43699, 24.655, 47.9592, 19.9489, 0,       0,       0,       0,
+       0, 0, 0, 0,       0,      2.20281, 2.46741, 24.4398, 17.8244, 39.8236, 13.242},
+      {0, 0, 0,       0,      56.7789, 43.2211, 0,       0,       0,      0, 0, 0,
+       0, 0, 11.6001, 22.861, 47.6905, 17.8484, 0,       0,       0,      0, 0, 1.00471,
+       0, 0, 0,       1.7485, 1.95555, 13.8623, 11.9821, 42.1631, 27.2837},
+      {0, 0, 0, 0,       56.8728, 43.1272, 0,       0,        0,       0,      0,
+       0, 0, 0, 6.01596, 26.1097, 49.8964, 17.7679, 0.210063, 0,       0,      0,
+       0, 0, 0, 0,       0,       2.53109, 5.06131, 19.2648,  25.4152, 36.075, 11.6525},
+      {0, 0, 0,       0,       51.8106, 48.1894, 0,        0,       0,      0, 0, 0,
+       0, 0, 5.59972, 19.2943, 54.1986, 20.6518, 0.255575, 0,       0,      0, 0, 0,
+       0, 0, 0,       6.78113, 7.8062,  15.5474, 31.9517,  25.3035, 12.6101},
+      {0, 0, 0, 0, 0.858694, 99.1413, 0,       0,       0,       0, 0,
+       0, 0, 0, 0, 0.188111, 99.8119, 0,       0,       0,       0, 0,
+       0, 0, 0, 0, 0,        6.26467, 82.2678, 10.3244, 1.14313, 0, 0},
+      {0, 0, 0, 0, 4.95132, 95.0487, 0, 0, 0, 0, 0,       0,      0,       0, 0, 0, 100,
+       0, 0, 0, 0, 0,       0,       0, 0, 0, 0, 5.61338, 81.416, 12.9706, 0, 0, 0},
+      {0, 0, 0, 0, 6.45045, 93.5496, 0,       0,       0, 0, 0,
+       0, 0, 0, 0, 0,       100,     0,       0,       0, 0, 0,
+       0, 0, 0, 0, 0,       5.79338, 80.1302, 14.0764, 0, 0, 0},
+      {0, 0, 0, 0, 3.19265, 96.8073, 0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       100,     0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       7.14492, 77.4834, 14.8963, 0.475413, 0, 0},
+      {0, 0, 0, 0, 2.91857, 97.0814, 0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       100,     0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       6.15287, 83.1738, 10.5024, 0.170916, 0, 0},
+      {0, 0, 0, 0, 2.7342, 97.2658, 0,       0,       0,       0, 0,
+       0, 0, 0, 0, 0,      100,     0,       0,       0,       0, 0,
+       0, 0, 0, 0, 0,      5.28806, 83.1876, 10.1264, 1.39793, 0, 0},
+      {0, 0, 0, 0, 8.13745, 91.8625, 0,       0,       0, 0, 0,
+       0, 0, 0, 0, 0,       100,     0,       0,       0, 0, 0,
+       0, 0, 0, 0, 0,       2.63515, 88.9699, 8.39498, 0, 0, 0},
+      {0, 0, 0, 0, 9.74746, 90.2525, 0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       100,     0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       6.45207, 80.3423, 12.6033, 0.602394, 0, 0},
+      {0, 0, 0, 0, 10.3765, 89.6235, 0, 0, 0, 0, 0,       0,       0,      0, 0, 0, 100,
+       0, 0, 0, 0, 0,       0,       0, 0, 0, 0, 5.73981, 82.4882, 11.772, 0, 0, 0},
+      {0, 0, 0, 0, 9.47275, 90.5273, 0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       100,     0,       0,       0,        0, 0,
+       0, 0, 0, 0, 0,       5.58698, 79.8493, 14.2933, 0.270383, 0, 0}};
```

Comment:
How were these numbers generated? Are these standard or is our implementation guaranteed to not have bugs?

Reply:
So my initial idea while writing the tests was to create test for functions which finds the mean of the features for example or for a function which finds the persistent values using a distribution. However, these functions were private. So I couldnt write tests for them since I cannot reach them. In the end I decided to use one of the pcd files in the project(bun0.pcd) well I checked and it gave me reasonable results so I took the generated features and used them in the test. These features generated using the example code. I know it is not the nicest way, do you have any other ideas how this could be done?

Reply:
Our options are pretty limited. If possible, I'd advise to make those functions protected. If the functions have some details that shouldn't be protected, but be private, then we can do the following:
```cpp
// Before
class A{
  void foo();
  // rest of the stuff
};

// After
namespace detail {
class B {
  public: void foo();  // or protected: foo
  // rest of A's stuff
};
}

class A: public B {
  private: B::foo;
  // everything else has the current visibility, no API/ABI change either :)
};

// Now we can test B since foo isn't private
// if we choose protected, then we inherit from B
// else it's pretty straightforward
```

Reply:
Will do this today! Thanks for the explanation